### DISCUSSION TITLE
Symfony Bundle - Form Type

### DISCUSSION THREAD

**Discussion by arnaud-ritti** [mosparo/mosparo#54] [Integrations]
Hi,

First at all, thank you for the job you are doing on mosparo, it's a friendly open-source and self hosted spam protection system.

I would be happy to give you my Symfony Bundle for mosparo integration. You can find it here [https://github.com/arnaud-ritti/mosparo-bundle](https://github.com/arnaud-ritti/mosparo-bundle).

Feel free to send me any update or improvement.

**Comment by zepich** [comment]
Hi @arnaud-ritti 

First: Thank you very much for your feedback. It's always great to get feedback!

Secondly: Thank you very much for your work with the bundle! By coincidence, I saw your repository last week after you created it (and when it still was empty), and I was very happy to see that you added some code some days later!

I've looked at the bundle, and as far as I can tell, everything looks very good. Your work has a very good code quality! I haven't tested it in a Symfony project yet, but I will do it soon.

## Recommend changes

1. I found one small "typo": In the README.md, you wrote, "How to integrate **re-captcha** in Symfony form:". I think `re-captcha` should be replaced with `mosparo`.

2. We strongly recommend not sending ignored data to mosparo. Let's say you have a password field in a form. The frontend script of mosparo will ignore this field and not validate it. But the `FormNormalizer` in your bundle will send all field data to mosparo for verification. That generally works since mosparo only checks the data it got from the frontend. But in terms of sensitive information, we recommend not doing that. Our approach to that is to ignore some field types when we gather all the form data and only send the not-ignored field data to mosparo for verification ([Example](https://github.com/mosparo/wordpress-plugin/blob/master/src/MosparoIntegration/Module/ElementorForm/MosparoField.php#L268)).

## Optional suggestions

1. Allow disabling the SSL verification by ENV variable
Especially for dev environments, you may not have a valid SSL certificate, so it should not be verified. The mosparo API Client has a flag to disable the SSL certificate verification ([Example](https://github.com/mosparo/wordpress-plugin/blob/master/src/MosparoIntegration/Helper/VerificationHelper.php#L40)).

2. Allow overriding of the connection settings (mosparo host, UUID, public and private key, verify SSL) in the field's options.
You can use mosparo in a contact and a login form. For the login form, you may want stronger security settings than in the contact form, so it would be good to have an option to set different connection details for login fields (for example) to another mosparo project.

3. After the verification is executed, it's essential to check that all verifiable fields got verified by mosparo. 
The problem is that the user could change an input field's `type` attribute and bypass the mosparo validation. By checking that all verifiable fields are verified, we can ensure that nobody bypasses the validation. In our integrations, we also check that all required fields got verified. You can find more about that in [our documentation](https://documentation.mosparo.io/docs/integration/bypass_protection#bypass-field-validation) or our [WordPress plugin (for example)](https://github.com/mosparo/wordpress-plugin/blob/master/src/MosparoIntegration/Module/ElementorForm/MosparoField.php#LL226C44-L226C44). _Disclaimer: We added the bypass page to the documentation yesterday, so you couldn't know that before._

4. We use a filter (WordPress), a callback (Django), or an Event (Mautic) to allow the developers to adjust the list of ignored and verifiable field types. Additionally, we offer an additional event to filter the prepared form data before they get sent to the mosparo API. I also recommend that for the Symfony bundle since you may have additional form types that should be ignored by mosparo. With an event, another developer could add (or remove) a field from the list of ignored or verifiable field types or adjust the prepared form data ([Example Field Types](https://github.com/mosparo/wordpress-plugin/blob/master/src/MosparoIntegration/Module/ElementorForm/MosparoField.php#L244) and [Example Form Data](https://github.com/mosparo/wordpress-plugin/blob/master/src/MosparoIntegration/Module/ElementorForm/MosparoField.php#L289)).

I'm very happy that you took the time to develop this bundle. I had it on my list for weeks (somehow, I forgot to do that in the last two years while we developed the project) but didn't have time to start developing it. When I saw that you'd developed it, and with which quality, I was very happy that I could take one item from my to-do list.

I added the bundle to our list of available integrations in the documentation yesterday and to the website today.

Again, thank you very much for your fantastic work!

Kind regards,

zepich

**Comment by arnaud-ritti** [comment]
Hi @zepich,

I'm glad you appreciate it. I've made a new release yesterday with your feedback. 

For the multi configuration, instead to edit the FormType, i handled it like doctrine connections into the configuration file (https://github.com/arnaud-ritti/mosparo-bundle#handle-multiples-configurations).

Let me know if you have other suggestions.

Kind regards,

Arnaud RITTI

**Comment by zepich** [comment]
Hi @arnaud-ritti 

Awesome! Thank you very much for your work and the new release! Perfect solution with the different projects in the configuration!

I tried to use it today in a project I'm working on. I had/have two (technical) issues:

### 1. `projects` missing in the configuration after installation

After I've installed the bundle with `composer require arnaud-ritti/mosparo-bundle` it generated the following error:

```
Symfony operations: 1 recipe (5f20984d4f02b7289bd35bc87e05b3c9)
  - Configuring arnaud-ritti/mosparo-bundle (>=1.2.0): From auto-generated recipe
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!  In ArrayNode.php line 226:
!!                                                                     
!!    The child config "projects" under "mosparo" must be configured.  
!!                                                                     
!!  
```

Can you adjust this so the installation will not end in this error?

### 2. Configuration cannot be validated

After manually creating the `mosparo.yaml` file and adding the configuration, I got an error that the instance_url is not a valid URL.

```
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!  In BaseNode.php line 469:
!!                                                                                 
!!    Invalid configuration for path "mosparo.projects.forms.instance_url": "inst  
!!    ance_url" is not a valid URL                                                 
!!                                                                                 
!!  
!!  In ExprBuilder.php line 187:
!!                                       
!!    "instance_url" is not a valid URL  
!!                                       
!!  
```

I've checked my ENV variable multiple times. It is a valid URL, and the ENV variable name is correct everywhere. I've then adjusted the Configuration tree builder and saw that the `$value` is an empty string (here: https://github.com/arnaud-ritti/mosparo-bundle/blob/main/src/DependencyInjection/Configuration.php#L67).

As far as I understand the issue (and according to [this comment](https://github.com/symfony/symfony/issues/33358#issuecomment-525352590)), the value is always empty when we're using ENV variables because they can change in the runtime. I guess the configuration should only check for the node's existence. When you initialize the client (here: https://github.com/arnaud-ritti/mosparo-bundle/blob/main/src/Services/MosparoClient.php#L21), we have to make sure that the URL and UUID are valid.

Please let me know if you have any questions or if I can help you with something.

Thank you again for your work, and have a nice day!

Kind regards,

zepich

**Comment by arnaud-ritti** [comment]
Hi @zepich,

Sorry for these issues. I fixed it in the last patch.

Kind regards,

Arnaud RITTI

**Comment by zepich** [comment]
Hi @arnaud-ritti 

That's no problem. Thank you very much for the new version!

It works perfectly!

Kind regards,

zepich
